File: LittleGame/LittleGame/Entity/ClientPlayer.cs
using LittleGame.States;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace LittleGame.Entity
{
    class ClientPlayer : ClientMapObject
    {
        private System.Windows.Forms.PictureBox pictureBox;
        private int id;
        private bool alive;
        public bool Alive { get => alive; }
        private bool dead;
        public bool Dead { get => dead; set => dead = value; }
        private int hp;

        private bool attack;
        public bool Attack { get => attack; set => attack = value; }
        private bool reload;
        public bool Reload { get => reload; set => reload = value; }
        private bool reloadDone;
        public bool ReloadDone { get => reloadDone; set => reloadDone = value; }
        public int bulletCount;
        private int maxBulletCount;

        private static System.Drawing.Bitmap[,] images =
        {
            {
                Properties.Resources.p1up,
                Properties.Resources.p1down,
                Properties.Resources.p1left,
                Properties.Resources.p1right,
                Properties.Resources.p1moveup,
                Properties.Resources.p1movedown,
                Properties.Resources.p1moveleft,
                Properties.Resources.p1moveright
            },
            {
                Properties.Resources.p2up,
                Properties.Resources.p2down,
                Properties.Resources.p2left,
                Properties.Resources.p2right,
                Properties.Resources.p2moveup,
                Properties.Resources.p2movedown,
                Properties.Resources.p2moveleft,
                Properties.Resources.p2moveright
            },
            {
                Properties.Resources.p3up,
                Properties.Resources.p3down,
                Properties.Resources.p3left,
                Properties.Resources.p3right,
                Properties.Resources.p3moveup,
                Properties.Resources.p3movedown,
                Properties.Resources.p3moveleft,
                Properties.Resources.p3moveright,
            },
            {
                Properties.Resources.p4up,
                Properties.Resources.p4down,
                Properties.Resources.p4left,
                Properties.Resources.p4right,
                Properties.Resources.p4moveup,
                Properties.Resources.p4movedown,
                Properties.Resources.p4moveleft,
                Properties.Resources.p4moveright
            }
        };

        public ClientPlayer(ClientPlayingState state, int id, int x, int y)
        {
            this.state = state;
            pictureBox = new System.Windows.Forms.PictureBox();
            this.id = id;
            this.alive = true;
            this.face = DOWN;
            this.point = new System.Drawing.Point(x, y);
            this.size = new System.Drawing.Size(50, 50);
            this.hp = 1;
            this.attack = false;
            this.reload = false;
            this.reloadDone = false;
            this.bulletCount = 6;
            this.maxBulletCount = 6;
            this.dead = false;

            LoadImage(images[this.id, this.face]);
            this.state.Controls.Add(pictureBox);
            pictureBox.BringToFront();
        }
        
        private void LoadImage(System.Drawing.Bitmap image)
        {
            this.pictureBox.Image = image;
            this.pictureBox.BackColor = System.Drawing.Color.Transparent;
            this.pictureBox.Location = point;
            this.pictureBox.Name = "pictureBox";
            this.pictureBox.Size = size;
            this.pictureBox.SizeMode = System.Windows.Forms.PictureBoxSizeMode.StretchImage;
            this.pictureBox.TabIndex = 0;
            this.pictureBox.TabStop = false;
        }

        public void SetPoint(int x, int y)
        {
            this.point.X = x;
            this.point.Y = y;
        }
        

        public override void UIUpdate()
        {
            if (alive)
            {
                if(dead)
                {
                    alive = false;
                    state.Controls.Remove(this.pictureBox);
                    return;
                }
                if (attack)
                {
                    state.clientBullets_List.Add(new ClientBullet(state, face, point.X + size.Width / 2, point.Y + size.Height / 2));
                    bulletCount--;
                    attack = false;
                }
                LoadImage(images[id, face]);
                pictureBox.BringToFront();
            }
        }
        public override void GameUpdate()
        {
            if (alive)
            {
                if (reloadDone)
                {
                    bulletCount = maxBulletCount;
                    reloadDone = false;
                    reload = false;
                }
            }
        }
    }
}

File: LittleGame_Android/app/src/main/java/com/example/littlegame/Entity/ClientPlayer.java
package com.example.littlegame.Entity;

import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.Rect;
import android.util.Log;

import com.example.littlegame.R;
import com.example.littlegame.States.ClientPlayingState;

public class ClientPlayer extends ClientMapObject{
    private Bitmap pictureBox;
    private int id;
    private boolean alive;
    public boolean getAlive() { return alive; }
    private int hp;

    private boolean attack;
    public void setAttack(boolean value) { attack = value; }
    private boolean reload;
    public boolean getReload() { return reload; }
    public void setReload(boolean value) { reload = value; }
    private boolean reloadDone;
    public boolean getReloadDone() { return reloadDone; }
    public void setReloadDone(boolean value) { reloadDone = value; }
    private int bulletCount;
    public int getBulletCount() { return bulletCount; }
    private int maxBulletCount;
    private int animateCount;

    private Bitmap[] bitmaps;


    public static final int ACTION_UP = 0;
    public static final int ACTION_DOWN = 1;
    public static final int ACTION_LEFT = 2;
    public static final int ACTION_RIGHT = 3;
    public static final int ACTION_MOVE_UP = 4;
    public static final int ACTION_MOVE_DOWN = 5;
    public static final int ACTION_MOVE_LEFT = 6;
    public static final int ACTION_MOVE_RIGHT = 7;

    public ClientPlayer(ClientPlayingState state, int id, int x, int y)
    {
        this.state = state;
        this.tileMap  = state.getTileMap();
        this.id = id;
        alive = true;
        direction = DOWN;
        action = ACTION_DOWN;
        animateCount = -1;

        px = x;
        py = y;
        dx = px;
        dy = py;
        width = 50;
        height = 50;
        vx = vy = 0;
        stepSize = 20;

        hp = 1;

        attack = false;
        reload = false;
        reloadDone = false;
        bulletCount = 6;
        maxBulletCount = 6;


        key_up = key_down = key_left = key_right = false;

        LoadImage();
    }

    private Rect getBitmapRect() {
        return new Rect(0,0, bitmaps[action].getWidth(), bitmaps[action].getHeight());
    }

    private void LoadImage()
    {
        if(id == 0) {
            bitmaps = new Bitmap[12];
            bitmaps[0] = BitmapFactory.decodeResource(state.getGsm().getDrawView().getResources(), R.drawable.p1up);
            bitmaps[1] = BitmapFactory.decodeResource(state.getGsm().getDrawView().getResources(), R.drawable.p1moveup_1);
            bitmaps[2] = BitmapFactory.decodeResource(state.getGsm().getDrawView().getResources(), R.drawable.p1moveup_2);
            bitmaps[3] = BitmapFactory.decodeResource(state.getGsm().getDrawView().getResources(), R.drawable.p1down);
            bitmaps[4] = BitmapFactory.decodeResource(state.getGsm().getDrawView().getResources(), R.drawable.p1movedown_1);
            bitmaps[5] = BitmapFactory.decodeResource(state.getGsm().getDrawView().getResources(), R.drawable.p1movedown_2);
            bitmaps[6] = BitmapFactory.decodeResource(state.getGsm().getDrawView().getResources(), R.drawable.p1left);
            bitmaps[7] = BitmapFactory.decodeResource(state.getGsm().getDrawView().getResources(), R.drawable.p1moveleft_1);
            bitmaps[8] = BitmapFactory.decodeResource(state.getGsm().getDrawView().getResources(), R.drawable.p1moveleft_2);
            bitmaps[9] = BitmapFactory.decodeResource(state.getGsm().getDrawView().getResources(), R.drawable.p1right);
            bitmaps[10] = BitmapFactory.decodeResource(state.getGsm().getDrawView().getResources(), R.drawable.p1moveright_1);
            bitmaps[11] = BitmapFactory.decodeResource(state.getGsm().getDrawView().getResources(), R.drawable.p1moveright_2);
        }
        else if(id == 1) {
            bitmaps = new Bitmap[8];
            bitmaps[0] = BitmapFactory.decodeResource(state.getGsm().getDrawView().getResources(), R.drawable.p2up);
            bitmaps[1] = BitmapFactory.decodeResource(state.getGsm().getDrawView().getResources(), R.drawable.p2down);
            bitmaps[2] = BitmapFactory.decodeResource(state.getGsm().getDrawView().getResources(), R.drawable.p2left);
            bitmaps[3] = BitmapFactory.decodeResource(state.getGsm().getDrawView().getResources(), R.drawable.p2right);
            bitmaps[4] = BitmapFactory.decodeResource(state.getGsm().getDrawView().getResources(), R.drawable.p2moveup);
            bitmaps[5] = BitmapFactory.decodeResource(state.getGsm().getDrawView().getResources(), R.drawable.p2movedown);
            bitmaps[6] = BitmapFactory.decodeResource(state.getGsm().getDrawView().getResources(), R.drawable.p2moveleft);
            bitmaps[7] = BitmapFactory.decodeResource(state.getGsm().getDrawView().getResources(), R.drawable.p2moveright);
        }
        else if(id == 2) {
            bitmaps = new Bitmap[8];
            bitmaps[0] = BitmapFactory.decodeResource(state.getGsm().getDrawView().getResources(), R.drawable.p3up);
            bitmaps[1] = BitmapFactory.decodeResource(state.getGsm().getDrawView().getResources(), R.drawable.p3down);
            bitmaps[2] = BitmapFactory.decodeResource(state.getGsm().getDrawView().getResources(), R.drawable.p3left);
            bitmaps[3] = BitmapFactory.decodeResource(state.getGsm().getDrawView().getResources(), R.drawable.p3right);
            bitmaps[4] = BitmapFactory.decodeResource(state.getGsm().getDrawView().getResources(), R.drawable.p3moveup);
            bitmaps[5] = BitmapFactory.decodeResource(state.getGsm().getDrawView().getResources(), R.drawable.p3movedown);
            bitmaps[6] = BitmapFactory.decodeResource(state.getGsm().getDrawView().getResources(), R.drawable.p3moveleft);
            bitmaps[7] = BitmapFactory.decodeResource(state.getGsm().getDrawView().getResources(), R.drawable.p3moveright);
        }
        else if(id == 3) {
            bitmaps = new Bitmap[8];
            bitmaps[0] = BitmapFactory.decodeResource(state.getGsm().getDrawView().getResources(), R.drawable.p4up);
            bitmaps[1] = BitmapFactory.decodeResource(state.getGsm().getDrawView().getResources(), R.drawable.p4down);
            bitmaps[2] = BitmapFactory.decodeResource(state.getGsm().getDrawView().getResources(), R.drawable.p4left);
            bitmaps[3] = BitmapFactory.decodeResource(state.getGsm().getDrawView().getResources(), R.drawable.p4right);
            bitmaps[4] = BitmapFactory.decodeResource(state.getGsm().getDrawView().getResources(), R.drawable.p4moveup);
            bitmaps[5] = BitmapFactory.decodeResource(state.getGsm().getDrawView().getResources(), R.drawable.p4movedown);
            bitmaps[6] = BitmapFactory.decodeResource(state.getGsm().getDrawView().getResources(), R.drawable.p4moveleft);
            bitmaps[7] = BitmapFactory.decodeResource(state.getGsm().getDrawView().getResources(), R.drawable.p4moveright);
        }
    }

    public void SetDestinationPoint(int x, int y)
    {
       dx = x;
       dy = y;
    }

    public void Hitted(int damage) {
        if (alive)
        {
            hp -= damage;
            if(hp <=0)
            {
                alive = false;
                state.setAliveNum(state.getAliveNum()-1);
            }
        }
    }

    public void Attack() {
        state.getBullets().add(new Bullet(state, id, direction, px + width / 2, py + height / 2));
        bulletCount--;
    }

    private void SetAction() {
        action = direction;
        if(id == 0) {
            if(vx != 0 || vy != 0) {
                animateCount = (animateCount+1) % 2;
                action = direction*3 + 1 + animateCount;
            }
            else {
                action = action = direction*3;
                animateCount = -1;
            }
        }
    }

    @Override
    public void Draw(Canvas canvas) {
        Paint paint = new Paint();
        SetAction();
        if (alive) {
            canvas.drawBitmap(bitmaps[action], getBitmapRect(), getDrawRectangle(), paint);
        }
    }

    private void MoveToDestination() {
        if(px != dx){
            vx = dx - px;
            px = dx;
        }
        else {
            vx = 0;
        }
        if(py != dy){
            vy = dy - py;
            py = dy;
        }
        else {
            vy = 0;
        }
    }

    @Override
    public void Update() {
        if (alive)
        {
            MoveToDestination();
            if(attack) {
                attack = false;
            }
            if (reloadDone)
            {
                bulletCount = maxBulletCount;
                reloadDone = false;
                reload = false;
            }
        }
    }
}
